### PR TITLE
Re-Entrancy tests

### PR DESCRIPTION
**Pull request by yrong**
For demonstration only, more context in discussion https://github.com/Snowfork/snowbridge/pull/927#discussion_r1305160460

### CODE REVIEW

File: contracts/test/mocks/HelloWorld.sol [hunk 1]
```diff
@@ -1,12 +1,59 @@
...
+        bytes memory call = abi.encodeCall(GatewayMock.setOperatingModePublic, abi.encode(params));
+
+        (bool success,) = gateway.call(call);
+
+        require(success, "Failed to reentrancy");
+
+        emit SaidHello("hacked");
+    }
+
+    function attack2(address gateway) public {
```
Comment (claravanstaden) [inline contracts/test/mocks/HelloWorld.sol:34]:
```suggestion
    function attemptReentrancyExploit(address gateway) public {
```

Maybe a more descriptive name? 😄

File: contracts/test/mocks/HelloWorld.sol [hunk 2]
```diff
@@ -1,12 +1,59 @@
...
+    uint8 internal reenter;
+
     event SaidHello(string indexed message);
 
     function sayHello(string memory _text) public {
         string memory fullMessage = string(abi.encodePacked("Hello there, ", _text));
         emit SaidHello(fullMessage);
     }
+
+    function attack(address gateway) public {
```
Comment (claravanstaden) [inline contracts/test/mocks/HelloWorld.sol:20]:
```suggestion
    function attemptExploitToSetGatewayToRejectOutboundMessages(address gateway) public {
```

Perhaps a more descriptive name?

File: contracts/test/mocks/HelloWorld.sol [hunk 3]
```diff
@@ -1,12 +1,59 @@
...
+        Gateway.SetOperatingModeParams memory params =
+            Gateway.SetOperatingModeParams({mode: OperatingMode.RejectingOutboundMessages});
+
+        bytes memory call = abi.encodeCall(GatewayMock.setOperatingModePublic, abi.encode(params));
+
+        (bool success,) = gateway.call(call);
+
+        require(success, "Failed to reentrancy");
+
+        emit SaidHello("hacked");
```
Comment (claravanstaden) [inline contracts/test/mocks/HelloWorld.sol:31]: So if this event is emitted, has the contract been successfully hacked? 🤔 The success or failure conditions aren't super clean for me here.

File: contracts/test/mocks/HelloWorld.sol [hunk 4]
```diff
@@ -1,12 +1,59 @@
...
+    function attack(address gateway) public {
+        console.log("gateway proxy address is: %s", gateway);
+        Gateway.SetOperatingModeParams memory params =
+            Gateway.SetOperatingModeParams({mode: OperatingMode.RejectingOutboundMessages});
+
+        bytes memory call = abi.encodeCall(GatewayMock.setOperatingModePublic, abi.encode(params));
+
+        (bool success,) = gateway.call(call);
+
+        require(success, "Failed to reentrancy");
```
Comment (claravanstaden) [inline contracts/test/mocks/HelloWorld.sol:29]: So if success == false, what happened? The re-etrancy attack failed? Or succeeded?

**Review by claravanstaden — COMMENTED** [review]
Hey @yrong! Thanks for adding these. I am struggling a bit to wrap my head around if these tests test that re-rentry attacks are not possible, or are they possible? When I read the code it seems to me that it tests that a re-entry attack can be done.

**Comment by yrong** [comment]
@claravanstaden This PR is just a temporary test and some assumptions here may not be correct.  I do think enabling arbitrary transact will open some back doors and increase the risk but let's focus on the basic flow first.  To not bother you I just closed it.

**Pull request closed without merging**